src/wire/causal_snapshot.rs

```diff
@@ -0,0 +1,200 @@
+//! A wire protocol for representing Modality probe causal snaphots
+
+use crate::{wire::le_bytes, ProbeId};
+use core::mem::size_of;
+use static_assertions::const_assert_eq;
+
+/// Everything that can go wrong when attempting to interpret a causal snaphot
+/// from the wire representation
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum CausalSnapshotWireError {
+    /// There weren't enough bytes for a full causal snapshot
+    MissingBytes,
+    /// The probe id didn't follow the rules for being
+    /// a valid Modality probe-specifying ProbeId
+    InvalidProbeId(u32),
+}
+
+/// A read/write wrapper around a causal snaphot buffer
+#[derive(Debug, Clone)]
+pub struct CausalSnapshot<T: AsRef<[u8]>> {
+    buffer: T,
+}
+
+const_assert_eq!(size_of::<crate::CausalSnapshot>(), field::REST.start);
+
+mod field {
+    type Field = ::core::ops::Range<usize>;
+    type Rest = ::core::ops::RangeFrom<usize>;
+
+    /// LogicalClock.id
+    pub const PROBE_ID: Field = 0..4;
+    /// LogicalClock.count
+    pub const COUNT: Field = 4..8;
+    /// Reserved field
+    pub const RESERVED_0: Field = 8..10;
+    /// Reserved field
+    pub const RESERVED_1: Field = 10..12;
+    /// Remaining bytes
+    pub const REST: Rest = 12..;
+}
+
+impl<T: AsRef<[u8]>> CausalSnapshot<T> {
+    /// Construct a causal snaphot from a byte buffer
+    pub fn new_unchecked(buffer: T) -> CausalSnapshot<T> {
+        CausalSnapshot { buffer }
+    }
+
+    /// Construct a causal snaphot from a byte buffer, with checks.
+    ///
+    /// A combination of:
+    /// * [new_unchecked](struct.CausalSnapshot.html#method.new_unchecked)
+    /// * [check_len](struct.CausalSnapshot.html#method.check_len)
+    pub fn new(buffer: T) -> Result<CausalSnapshot<T>, CausalSnapshotWireError> {
+        let r = Self::new_unchecked(buffer);
+        r.check_len()?;
+        Ok(r)
+    }
+
+    /// Ensure that no accessor method will panic if called.
+    ///
+    /// Returns `Err(CausalSnapshotWireError::MissingBytes)` if the buffer
+    /// is too short.
+    pub fn check_len(&self) -> Result<(), CausalSnapshotWireError> {
+        let len = self.buffer.as_ref().len();
+        if len < field::REST.start {
+            Err(CausalSnapshotWireError::MissingBytes)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Consumes the causal snaphot, returning the underlying buffer
+    pub fn into_inner(self) -> T {
+        self.buffer
+    }
+
+    /// Return the length of a buffer required to hold a causal snaphot
+    pub fn buffer_len() -> usize {
```

Comment:
This method name seems confusing, because it doesn't actually return the buffer's len, it returns an expectation about what a buffer's len should at least be. A careful reader will note the lack of a `self` parameter, but perhaps for clarity we could make it a `const` function with a slightly more explicit name. Maybe `min_buffer_len` ?

Reply:
The name change makes sense to me. 
I held off on making the associated methods `const` to avoid `feature(const_fn)`, I think it's still unstable:
> error[E0723]: trait bounds other than `Sized` on const fn parameters are unstable